src/types.mligo

```diff
@@ -146,6 +146,17 @@ type governance_token =
   ; token_id : token_id
   }
 
+type period_frozen_history =
+  { current : nat
+  ; last : nat
+  ; updated_period : nat
+  }
+
+type quorum_threshold_with_period =
+  { quorum_threshold : quorum_threshold
+  ; updated_period : nat
+  }
```

Comment:
I think this representation can be simplified a lot and as a result the code would be simpler as well.

AFAIU for a given period `x` all the information needed to calculate the `quorum_threshold` are:
- constants
- the `quorum_threshold` of period `x - 1` (aka `old_quorum`)
- the `total_supply` of frozen tokens at period `x - 1`

(the exception is period `0`, but that's not a problem as all the info is known at the beginning)

So what I think we should so is track all we need in a single type:
```ocaml
type quorum_threshold_at_stage =
  { quorum_threshold : quorum_threshold
  ; frozen_supply: nat
  ; last_updated_period : nat
  }
```
To keep the information in it correct (and enough) we need to update this lazily in its entirety whenever:
- we change the `total_supply` of frozen tokens (in `freeze`/`unfreeze`)
- we need the `quorum_threshold`

I'll write more comments about the updates below.